src/Features/Core/Portable/Formatting/FormattingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Formatting
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal class FormattingDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public FormattingDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.FormattingDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_analyzer_title), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_analyzer_message), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SyntaxAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxTreeAction(AnalyzeSyntaxTree);
+
+        private void AnalyzeSyntaxTree(SyntaxTreeAnalysisContext context)
+        {
+            if (!(context.Options is WorkspaceAnalyzerOptions workspaceAnalyzerOptions))
+            {
+                return;
+            }
+
+            var tree = context.Tree;
+            var cancellationToken = context.CancellationToken;
+
+            var options = context.Options.GetDocumentOptionSetAsync(tree, cancellationToken).GetAwaiter().GetResult();
+            if (options == null)
+            {
+                return;
+            }
+
+            var workspace = workspaceAnalyzerOptions.Services.Workspace;
+            var oldText = tree.GetText(cancellationToken);
+            var formattingChanges = Formatter.GetFormattedTextChanges(tree.GetRoot(cancellationToken), workspace, options, cancellationToken);
+
+            // formattingChanges could include changes that impact a larger section of the original document than
+            // necessary. Before reporting diagnostics, process the changes to minimize the span of individual
+            // diagnostics.
+            foreach (var formattingChange in formattingChanges)
```

Comment:
>foreach (var formattingChange in formattingChanges) [](start = 12, length = 51)

Is the code in this entire loop identical between here and the analyzer implementation in the CodeStyle layer? If so, consider moving it into a shared helper that is linked into both the projects.

Reply:
👍

Reply:
Extracted to `FormattingAnalyzerHelper`.